### ISSUE TITLE
Varying batch size gives different results when dynamic tracing is enabled

### ISSUE THREAD

**Issue by analyticsbot** [aws-neuron/aws-neuron-sdk#710]
Hello all, we are working to trace a Pytorch model using `torch.neuron.trace`. As I understand the 0th dimension can be dynamic. I have the following operations inside the code (context below why this piece is mentioned)

``` val1    = torch.mean(variable, dim=0, keepdim=True)  
       val2, _ = torch.max(variable, dim=0, keepdim=True)           
        final_value = torch.cat([val1,val2], dim=1)             
```
What we are finding is that the output from the model is heavily dependent on the batch size of the input. So for example, if the model was traced with batch size of 1000 (input size is of dimension (1000,72)) then the output is what is expected (compared to untraced model, output varies between 0 and 1). But if the model was traced with batch size of 10 then the output is much lower than what is expected. If the model was traced with batch size of 10000, then the output is higher than expected (almost 1).

I tried to Google around and found that some operations such as mean and sum might behave differently (code shared above) inside a neuron trace and tried finding ways to bypass that but it is not solving the issue.

Unfortunately, I can't share the code here. Anyone has any thoughts what can I do to fix this issue?

**Comment by analyticsbot** [comment]
Btw, the tracing happens successfully with no errors.
The model output before the tracing exactly matches the expected value, only with the traced model these anomalies are being observed.

Also, tried looking into the https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/appnotes/torch-neuron/torch-neuron-dataparallel-app-note.html
but it didn't help.

Came across bucketing and padding but don't think it will solve the issue.

**Comment by aws-taylor** [comment]
Hello @analyticsbot,

 This may be difficult to hunt down without any ability to reproduce the issue. Are you passing `dynamic_batch_size=True` as part of your call to `torch.neuron.trace()`? 
 
 Documentation: https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/arch/neuron-features/neuroncore-batching.html

**Comment by analyticsbot** [comment]
Hello @aws-taylor - thank you for your reply.
Yes, I am passing `dynamic_batch_size=True` as part of my call to `torch.neuron.trace()`.

To explain a bit more, our model used resnet50 as the backbone but does more operations on top of it to achieve the final result. Our model is currently split in two stages. Each stage's input is different. For stage 1, we split the image into different patches and pass the patches into the model.

For stage 2, we take the output of stage 1 and use that as input for the stage 2. 

The reason for doing this - since the BATCH dimension is the only dimension that can change and since stage 1 and stage 2 operation at different dimensions, we decide to create this stage 1 and stage 2.

Stage 1 operates at [Num Patches, Num Channels, X Dim, Y Dim] -> [Num Patches, Embedding Size]
Stage 2 operates at [Num Patches, Embedding Size] -> Probability

Num Patches is divided on the input image size. As I said earlier, if we trace with [Num Patches, Num Channels, X Dim, Y Dim] & [Num Patches, Embedding Size] for the two stages respectively, then we get similar result after tracing (before tracing exactly same result). However if the Num Patches is different than what we traced, we get different output.

One thing that I can think of - in normal operation batch size is for different images and for each image is one item in that batch. In this case, the image itself is split into patches and is used as the batch dimension. Taken from here [https://github.com/aws-neuron/aws-neuron-sdk/blob/master/src/examples/pytorch/resnet50.ipynb]

> Run inference with dynamic batch sizes
> Batch size has a direct impact on model performance. The Inferentia chip is optimized to run with small batch sizes. This means that a Neuron compiled model can outperform a GPU model, even if running single digit batch sizes.
> 

> As a general best practice, we recommend optimizing your model's throughput by compiling the model with a small batch size and gradually increasing it to find the peak throughput on Inferentia.
> 
> Dynamic batching is a feature that allows you to use tensor batch sizes that the Neuron model was not originally compiled against. This is necessary because the underlying Inferentia hardware will always execute inferences with the batch size used during compilation. Fixed batch size execution allows tuning the input batch size for optimal performance. For example, batch size 1 may be best suited for an ultra-low latency on-demand inference application, while batch size > 1 can be used to maximize throughput for offline inferencing. Dynamic batching is implemented by slicing large input tensors into chunks that match the batch size used during the torch.neuron.trace compilation call.
> 
> The torch.neuron.DataParallel class automatically enables dynamic batching on eligible models. This allows us to run inference in applications that have inputs with a variable batch size without needing to recompile the model.
> 
> In the following example, we use the same torch.neuron.DataParallel module to run inference using several different batch sizes. Notice that latency increases consistently as the batch size increases. Throughput increases as well, up until a certain point where the input size becomes too large to be efficient.
> 

Specifically `This is necessary because the underlying Inferentia hardware will always execute inferences with the batch size used during compilation.Dynamic batching is implemented by slicing large input tensors into chunks that match the batch size used during the torch.neuron.trace compilation call.`
But we need to look at all patches that we create to arrive at the final value.

Here is how we trace the two stages.

```
import torch 
import torch_neuron ## using 1.13.1+cu117 1.13.1.2.7.1.0
# Stage 1 Tracing 
input_stage1 = torch.rand([Num Patches, Num Channels, X Dim, Y Dim]) 
stage1_traced_model = torch.neuron.trace(stage1_model, example_inputs=input_stage1, dynamic_batch_size=True)
# saving scripted model 
traced_model.save(model_path) # we also tried this torch.jit.save(stage1_traced_model, model_path) 

## JIT Trace
stage1_jit_model = torch.jit.script(stage1_model) ## output from jit trace match
torch.jit.save(stage1_jit_model, model_path) 

# # Stage 2 
torch._C._jit_set_texpr_fuser_enabled(False)  # tried with and without this
input_stage2 = torch.rand([Num Patches, Embedding Size]) 
stage2_traced_model = torch.neuron.trace(stage2_model, example_inputs=input_stage2, dynamic_batch_size=True, compiler_args=['--fast-math', 'none']) # tried without the arguments compiler_args=['--fast-math', 'none']
stage2_traced_model.save(model_path) 

## JIT Trace
stage2_jit_model = torch.jit.script(stage2_model) ## output from jit trace match
torch.jit.save(stage2_jit_model, model_path) 

```

Sample output
``` 
INFO:Neuron:All operators are compiled by neuron-cc (this does not guarantee that neuron-cc will successfully compile) 
INFO:Neuron:Number of arithmetic operators (pre-compilation) before = 180, fused = 180, percent fused = 100.0%
WARNING:tensorflow:From torch_neuron/ops/aten.py:2387: where (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version. Instructions for updating: Use tf.where in 2.0, which has the same broadcast rule as np.where
INFO:Neuron:Compiling function _NeuronGraph$560 with neuron-cc INFO:Neuron:Compiling with command line: '/home/model-server/tmp/tmph2yzhmjc/graph_def.pb --framework TENSORFLOW --pipeline compile SaveTemps --output /tmp/tmph2yzhmjc/graph_def.neff --io-config {"inputs": {"tensor:0": [[Num Patches, Num Channels, X Dim, Y Dim], "float32"]}, "outputs": ["aten_reshape/Reshape:0"]} --verbose 35

INFO:Neuron:Number of arithmetic operators (post-compilation) before = 180, compiled = 180, percent compiled = 100.0% 
INFO:Neuron:The neuron partitioner created 1 sub-graphs 
INFO:Neuron:Neuron successfully compiled 1 sub-graphs, Total fused subgraphs = 1, Percent of model sub-graphs successfully compiled = 100.0% 
INFO:Neuron:Compiled these operators (and operator counts) to Neuron: 
INFO:Neuron: => aten::Int: 1 
INFO:Neuron: => aten::_convolution: 55 
INFO:Neuron: => aten::add_: 16 
INFO:Neuron: => aten::batch_norm: 54 
INFO:Neuron: => aten::cat: 1 
INFO:Neuron: => aten::relu_: 50 
INFO:Neuron: => aten::reshape: 1 
INFO:Neuron: => aten::size: 1 
INFO:Neuron: => aten::slice: 1

```

**Comment by analyticsbot** [comment]
Hi @aws-taylor - did the response above give you more context on what we are trying to achieve and where the potential problem might be.
Another question - does neuron trace support print to screen and loggers?

**Comment by Poojaakkaladevi** [comment]
Hi @aws-taylor,

I'm trying to trace the resnet model on inf1 instance with tensor size (1,3,224,224) as shown in AWS Neuron Documentation.
[[https://awsdocs-neuron.readthedocs-hosted.com/en/latest/src/examples/pytorch/resnet50.html](url)]

When I tried to run inference on the input size tensor(2,3,224,224) with the traced models--> (with and without dynamic_batch_size argument as True, just to see if this argument makes any difference)

- the outputs from both of the traced models don't match with the actuals (untraced model output) but are similar with it. Is this expected?
- But the outputs from torch.jit.trace() exactly match with the actuals 

Below is the picture of the code I'm using:

- In the First cell, I'm tracing the neural network model passing dynamic_batch_size argument as True once and as False once, also tracing the model with torch.jit.trace()
- In the second cell, Printing out the outputs from all four models
![image](https://github.com/aws-neuron/aws-neuron-sdk/assets/34411859/34d872e9-99c4-4003-b8a0-24c79d46369e)

**Comment by jluntamazon** [comment]
> the outputs from both of the traced models don’t match with the actuals (untraced model output) but are similar with it. Is this expected?

Some accuracy differences are expected since we use mixed precision casting by default. On `inf1` the default is to use `bf16` autocasting. This will produce the probability deviations you see in the example but this is usually not significant enough to change classification results.

If different precision is required, we have a set of compiler options that can be used to tune accuracy/performance.

See the `--fast-math` flag in the compiler reference: https://awsdocs-neuron.readthedocs-hosted.com/en/latest/compiler/neuron-cc/command-line-reference.html

Mixed precision guide:  https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/appnotes/neuron-cc/mixed-precision.html

For resnet50, the **most accurate** option is likely to use `fp16` matrix multiplication, native data types for other operations, and the accurate relayout flag. This can be enabled with:
```
torch.neuron.trace(model, image, compiler_args='--fp32-cast matmult-fp16 --fast-math no-fast-relayout')
```
Note that this may be **slower** than the default options, so it is important to try the different casting options to see what is fits best to your application.

**Comment by micwade-aws** [comment]
@analyticsbot - please reach out at aws-neuron-support@amazon.com to work out a way to exchange code privately, if needed.
